indexing-service/src/main/java/org/apache/druid/indexing/seekablestream/supervisor/autoscaler/CostBasedAutoScaler.java

```diff
@@ -0,0 +1,275 @@
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+package org.apache.druid.indexing.seekablestream.supervisor.autoscaler;
+
+import org.apache.druid.indexing.overlord.supervisor.SupervisorSpec;
+import org.apache.druid.indexing.overlord.supervisor.autoscaler.LagStats;
+import org.apache.druid.indexing.overlord.supervisor.autoscaler.SupervisorTaskAutoScaler;
+import org.apache.druid.indexing.seekablestream.supervisor.SeekableStreamSupervisor;
+import org.apache.druid.java.util.common.StringUtils;
+import org.apache.druid.java.util.common.concurrent.Execs;
+import org.apache.druid.java.util.emitter.EmittingLogger;
+import org.apache.druid.java.util.emitter.service.ServiceEmitter;
+import org.apache.druid.java.util.emitter.service.ServiceMetricEvent;
+import org.apache.druid.query.DruidMetrics;
+
+import java.util.ArrayList;
+import java.util.Arrays;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ScheduledExecutorService;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicReference;
+
+/**
+ * Cost-based auto-scaler for seekable stream supervisors.
+ * Uses a cost function combining lag and idle time metrics to determine optimal task counts.
+ * Task counts are selected from predefined values (not arbitrary factors).
+ * Scale-up happens incrementally, scale-down only during task rollover.
+ */
+public class CostBasedAutoScaler implements SupervisorTaskAutoScaler
+{
+  private static final EmittingLogger log = new EmittingLogger(CostBasedAutoScaler.class);
+
+  private static final int SCALE_FACTOR_DISCRETE_DISTANCE = 2;
+  public static final String OPTIMAL_TASK_COUNT_METRIC = "task/autoScaler/costBased/optimalTaskCount";
+
+  private static final Map<Integer, int[]> FACTORS_CACHE = new LinkedHashMap<>();
+  private static final int FACTORS_CACHE_MAX_SIZE = 10; // Enough for most scenarios, almost capacity * loadFactor.
+
+  private final String supervisorId;
+  private final SeekableStreamSupervisor supervisor;
+  private final ServiceEmitter emitter;
+  private final SupervisorSpec spec;
+  private final CostBasedAutoScalerConfig config;
+  private final ServiceMetricEvent.Builder metricBuilder;
+  /**
+   * Atomic reference to CostMetrics object. All operations must be performed
+   * with sequentially consistent semantics (volatile reads/writes).
+   * However, it may be fine-tuned with acquire/release semantics,
+   * but requires careful reasoning about correctness.
+   */
```

Comment:
We can avoid this javadoc as it does not add any new info. In fact, I think we can get rid of this field altogether.

Reply:
Good catch!